src/mgmt/srv_query.c

```diff
@@ -293,14 +294,76 @@ bio_query_dev_list(void *arg)
 	return 0;
 }
 
+static int
+copy_str2ctrlr(char **dst, const char *src)
+{
+	int len;
+
+	D_ASSERT(src != NULL);
+	D_ASSERT(dst != NULL);
+
+	if ((*dst != NULL) && (strnlen(*dst, NVME_DETAIL_BUFLEN) != 0)) {
+		D_ERROR("attempting to copy to non-empty destination");
+		return -DER_INVAL;
+	}
+
+	len = strnlen(src, NVME_DETAIL_BUFLEN);
+	if (len == NVME_DETAIL_BUFLEN) {
+		D_ERROR("src buf too large");
+		return -DER_INVAL;
+	}
+
+	D_ALLOC(*dst, len + 1);
+	if (*dst == NULL)
+		return -DER_NOMEM;
+
+	if (copy_ascii(*dst, len + 1, src, len) != 0) {
+		D_ERROR("copy_ascii: '%s'\n", src);
+		return -DER_INVAL;
+	}
+
+	return 0;
+}
+
+static int
+add_ctrlr_details(Ctl__NvmeController *ctrlr, struct bio_dev_info *dev_info)
+{
+	int rc;
+
+	rc = copy_str2ctrlr(&ctrlr->pci_addr, dev_info->bdi_traddr);
+	if (rc != 0)
+		return rc;
+	rc = copy_str2ctrlr(&ctrlr->model, dev_info->bdi_ctrlr->model);
+	if (rc != 0)
+		return rc;
```

Comment:
If we're aborting after some values were already copied, do we need to free the allocated values before returning?

Reply:
`ds_mgmt_smd_free_dev()` is called on each array element (device) in case of non-zero rc, this will free all Fields of the ctrlr object, regardless of when they were populated.